gcovr/__main__.py

```diff
@@ -218,8 +222,19 @@ def main(args=None):
                 "Bad --object-directory option.\n"
                 "\tThe specified directory does not exist.")
             sys.exit(1)
+        if options.use_canonical_paths:
+            canonical_objdir = os.path.realpath(options.objdir)
+            if canonical_objdir != options.objdir:
+                options.objdir = canonical_objdir
+                logger.msg(f"--object-directory has been normalized to {options.objdir}.")
```

Comment:
Can be moved to a function which depends on the option.
E.g. define at the start:
```python
        canonical_path = None
        if options.use_canonical_paths:
            def handle_path(dir, option=None):
                canonical_dir = os.path.realpath(dir)
                if canonical_dir != dir:
                    dir = canonical_dir
                    logger.msg(f"{dir if option == None else option} has been normalized to {dir}.")
                return dir
```
Usage of function:
```suggestion
        if canonical_path:
            options.objdir = canonical_path(options.objdir, "--object-dir")
```

Reply:
Even simpler would be a function that already knows whether this feature is enabled, so kind of like the “null object pattern”:

```python
if options.use_canonical_path:
  def canonical_path(path: str, option: str) -> str:
    canonical = os.path.realpath(path)
    if canonical != path:
      logger.msg(f"{option} has been normalized to {canonical}")
    return canonical
else:
  def canonical_path(path: str, option: str) -> str:
    return path
```

Reply:
That's even better.
What do you think about the option `--path_handling` to also handle the case of #420?

Reply:
I've adjusted this in the code as proposed by @latk.